### PR TITLE
Code quality improvements.

### PR DESCRIPTION
**Pull request by cgranade**
This PR makes code quality improvements and removes dependencies on private operations and functions within the standard library.

### CODE REVIEW

File: Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs [hunk 1]
```diff
@@ -41,25 +41,23 @@ namespace Microsoft.Quantum.Research.Chemistry {
...
-            let (minInt, bitStingApplyCNOT) = _DeltaParityCNOTbitstring_(prevFermionicTerm, nextFermionicTerm);
+    operation ApplyDeltaParity(
+        prevFermionicTerm : Int[],
+        nextFermionicTerm : Int[],
+        aux : Qubit,
+        qubits : Qubit[]
+    )
+    : Unit is Adj + Ctl {
+        // Circuit with cancellation of neighbouring CNOTS
+        let (minInt, bitStingApplyCNOT) = _DeltaParityCNOTbitstring_(prevFermionicTerm, nextFermionicTerm);
```
Comment (alan-geller) [inline Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs]:
```suggestion
        let (minInt, bitStringApplyCNOT) = _DeltaParityCNOTbitstring_(prevFermionicTerm, nextFermionicTerm);
```

File: Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs [hunk 2]
```diff
@@ -41,25 +41,23 @@ namespace Microsoft.Quantum.Research.Chemistry {
...
+        
+        for (idx in 0 .. Length(bitStingApplyCNOT) - 1) {
```
Comment (alan-geller) [inline Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs]:
```suggestion
        for (idx in 0 .. Length(bitStringApplyCNOT) - 1) {
```

File: Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs [hunk 3]
```diff
@@ -41,25 +41,23 @@ namespace Microsoft.Quantum.Research.Chemistry {
...
             
-            for (idx in 0 .. Length(bitStingApplyCNOT) - 1) {
-                
-                if (bitStingApplyCNOT[idx] == true) {
-                    CNOT(qubits[idx + minInt], aux);
-                }
+            if (bitStingApplyCNOT[idx] == true) {
```
Comment (alan-geller) [inline Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs]:
```suggestion
            if (bitStringApplyCNOT[idx] == true) {
```

File: Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs (hunk 1 above)
Comment (alan-geller) [reply Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs]: Also, shouldn't `_DeltaParityCNOTbitstring_` get the trailing _ removed?

File: Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs [hunk 4]
```diff
@@ -241,44 +208,37 @@ namespace Microsoft.Quantum.Research.Chemistry {
...
-            let y = _Ybasis_(_);
-            let ops = [[x, x], [y, y]];
-            let op0 = _JWOptimizedZ_(angle, parityQubit, _);
-            let op1 = ApplyWithCA(CNOTChainTarget([qubitP], _), op0, _);
-            
-            for (idxOp in 0 .. Length(ops) - 1) {
-                let op2 = ApplyWithCA(_ApplyBasisChange_(ops[idxOp], [qubitP], _), op1, _);
-                op2(qubitQ);
-            }
+    operation _JWOptimizedHpqTerm__ (
```
Comment (alan-geller) [inline Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs]: I kind of hate the underscores here, although I don't have a better suggestion off-hand.

File: Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs [hunk 5]
```diff
@@ -241,44 +208,37 @@ namespace Microsoft.Quantum.Research.Chemistry {
...
+    )
+    : Unit is Adj + Ctl {
+        let ((idxTermType, coeff), idxFermions) = term!;
+        let angle = (1.0 * coeff[0]) * stepSize;
+        let qubitP = qubits[idxFermions[0]];
+        let qubitQ = qubits[idxFermions[1]];
+        let qubitsPQ = Subarray(idxFermions[0 .. 1], qubits);
+        let x = _Xbasis;
+        let y = _Ybasis;
+        let ops = [[x, x], [y, y]];
```
Comment (alan-geller) [inline Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs]:
```suggestion
        let ops = [[_Xbasis, _Xbasis], [_Ybasis, _Ybasis]];
```
This lets you delete the two previous lines, also.

File: Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs (hunk 1 above)
Comment (cgranade) [reply Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs]: Good catch, fixing now.

File: Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs (hunk 4 above)
Comment (cgranade) [reply Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs]: I had the same thought as well; I think this is a [code smell](https://en.wikipedia.org/wiki/Code_smell) for a place where new Q# features can help us organize code better than we could with the Q# version that was used to write this code initially.

Comment (cgranade) [reply Research/src/Chemistry/JordanWignerOptimizedEvolutionSet.qs]: Regardless, though, I think I'll leave this for a future PR if that's alright. There's always more improvements that can be made, to be sure.

**Pull request merged**